### ISSUE TITLE
ConnectionTimeout not working on a server that is not responding

### ISSUE THREAD

**Issue by apoguy** [brettwooldridge/HikariCP#1289]
#### Environment
```
HikariCP version: x.x.x
JDK version     : 1.8.0_191
Database        : Oracle
Driver version  : 12.2.0
```
Found that the connectionTimeout property is not making any effect when you connect to a TCP that doesn't reply (in the latest versions (tested in 3.2.0)).

Trying to connect to a SocketTest Server, without sending any reply, the application hangs on `2018-12-07 12:45:39,330 main                           INFO  com.zaxxer.hikari.HikariDataSource                                                                                              []                            HikariPool-1 - Starting...` regardless of the time set in connectionTimeout.

In previous versions (working in 2.4.2) after connectionTimeout, yopu get the exception:

```
2018-12-07 12:48:15,804 main                           INFO  com.zaxxer.hikari.HikariDataSource                                                                                              []                            HikariPool-0 - is starting.
2018-12-07 12:48:45,819 Hikari housekeeper (pool HikariPool-0) DEBUG com.zaxxer.hikari.pool.HikariPool                                                                                               []                            Before cleanup   pool HikariPool-0 stats (total=0, active=0, idle=0, waiting=1)
2018-12-07 12:48:45,821 Hikari housekeeper (pool HikariPool-0) DEBUG com.zaxxer.hikari.pool.HikariPool                                                                                               []                            After cleanup    pool HikariPool-0 stats (total=0, active=0, idle=0, waiting=1)
2018-12-07 12:48:45,822 main                           DEBUG com.zaxxer.hikari.pool.HikariPool                                                                                               []                            Timeout failure  pool HikariPool-0 stats (total=0, active=0, idle=0, waiting=0)
2018-12-07 12:48:45,824 main                           INFO  com.zaxxer.hikari.pool.HikariPool                                                                                               []                            HikariPool-0 - is closing down.
....
```
Was this removed or is it an expected behavior?

**Comment by brettwooldridge** [comment]
@apoguy Our ``connectionTimeout`` unit tests are still passing, do you have a self-contained reproduction harness (code) that I can run to investigate?

**Comment by apoguy** [comment]
@brettwooldridge not really. The only tests we made are with our app, by switching the jar. We are using SocketTest (http://sockettest.sourceforge.net/) to open a dummy server and trying Hikari to connect to that server. SocketTest reports the connection opening, but doesn't send any response. If we try a live database with wrong parameters (user, password, etc.) we do get the error, while with SocketTest nothing happens. With the old version it works as expected.

If you need it, I can post pieces of the code used to get the configuration from Hikari file, etc.

**Comment by brettwooldridge** [comment]
@apoguy Can you post your full HikariCP configuration?

**Comment by apoguy** [comment]
@brettwooldridge sure thing.

My hikari.properties file is completely empty.

And therefore it's using the default config, here's the log output:

```
2018-12-11 11:41:50,970 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            HikariPool-1 - configuration:
2018-12-11 11:41:50,976 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            allowPoolSuspension.............false
2018-12-11 11:41:50,978 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            autoCommit......................true
2018-12-11 11:41:50,979 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            catalog.........................none
2018-12-11 11:41:50,979 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            connectionInitSql...............none
2018-12-11 11:41:50,980 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            connectionTestQuery.............none
2018-12-11 11:41:50,981 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            connectionTimeout...............30000
2018-12-11 11:41:50,983 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            dataSource......................none
2018-12-11 11:41:50,985 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            dataSourceClassName.............none
2018-12-11 11:41:50,995 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            dataSourceJNDI..................none
2018-12-11 11:41:51,000 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            dataSourceProperties............{password=<masked>}
2018-12-11 11:41:51,001 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            driverClassName................."oracle.jdbc.OracleDriver"
2018-12-11 11:41:51,005 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            healthCheckProperties...........{}
2018-12-11 11:41:51,012 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            healthCheckRegistry.............none
2018-12-11 11:41:51,013 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            idleTimeout.....................600000
2018-12-11 11:41:51,015 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            initializationFailTimeout.......1
2018-12-11 11:41:51,016 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            isolateInternalQueries..........false
2018-12-11 11:41:51,020 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            jdbcUrl.........................jdbc:oracle:thin:@localhost:9899:non_existent
2018-12-11 11:41:51,031 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            leakDetectionThreshold..........0
2018-12-11 11:41:51,032 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            maxLifetime.....................1800000
2018-12-11 11:41:51,032 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            maximumPoolSize.................10
2018-12-11 11:41:51,034 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            metricRegistry..................none
2018-12-11 11:41:51,036 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            metricsTrackerFactory...........none
2018-12-11 11:41:51,037 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            minimumIdle.....................10
2018-12-11 11:41:51,043 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            password........................<masked>
2018-12-11 11:41:51,044 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            poolName........................"HikariPool-1"
2018-12-11 11:41:51,046 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            readOnly........................false
2018-12-11 11:41:51,046 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            registerMbeans..................false
2018-12-11 11:41:51,047 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            scheduledExecutor...............none
2018-12-11 11:41:51,048 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            schema..........................none
2018-12-11 11:41:51,051 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            threadFactory...................internal
2018-12-11 11:41:51,054 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            transactionIsolation............default
2018-12-11 11:41:51,058 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            username........................"TEST"
2018-12-11 11:41:51,059 main                           DEBUG com.zaxxer.hikari.HikariConfig                                                                                                  []                            validationTimeout...............5000
2018-12-11 11:41:51,067 main                           INFO  com.zaxxer.hikari.HikariDataSource                                                                                              []                            HikariPool-1 - Starting...
```

and it just stays there forever.

just switching to the old jar works.

In SocketTest I see the connection:

```
> Server Started on Port: 9899
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> New Client: 127.0.0.1
[SOME DATA THAT CAN'T BE PASTED](DESCRIPTION=(CONNECT_DATA=(SID=non_existent)(CID=(PROGRAM=JDBC Thin Client)(HOST=__jdbc__)(USER=MyPc)))(ADDRESS=(PROTOCOL=TCP)(HOST=127.0.0.1)(PORT=9899)))
```

**Comment by brettwooldridge** [comment]
@apoguy I believe the change in behavior is due to the change in how ``initializationFailTimeout`` works.  Basically, I think you want to set ``initializationFailTimeout=-1``.  Have a look at the [docs](https://github.com/brettwooldridge/HikariCP#infrequently-used).

Another, possibly better, option is setting the ``oracle.net.CONNECT_TIMEOUT`` driver property (value in milliseconds).  This sets a TCP-level timeout that could prevent hung connections.

Regarding the change in behavior vs. HikariCP v2.4.x, the current pool behavior is actually better, and more flexible, provided that the driver has been properly configured.

If the ``oracle.net.CONNECT_TIMEOUT`` solves the issue, I would probably recommend leaving the ``initializationFailTimeout`` at its current default.

**Comment by musiKk** [comment]
We were also bitten by this recently. In our particular case the error condition happened when using HikariCP for a connection to an AWS RDS database. AWS provides the ability to fail over to another availability zone (AZ). In this case, a new instance is spun up in the other AZ, the DNS record for the database connection is changed to point to the new instance and the original AZ serves as the new secondary zone.

This is all well and good but for some weird reason, the connections to the original instance are never purged from the pool. Running with default timeout settings (and HikariCP 2.5.1) even after almost two hours the connections were still blocking with no exception being thrown.

I recreated the issue in a test environment (using version 2.7.9; going to 3.x is currently not an option for us). It seems in this particular case of a non-responding server, the `connectionTimeout` does not apply. Note that in the "normal" case of a database that is shut down, the exception says

```
java.sql.SQLTransientConnectionException: HikariPool - 
    Connection is not available, request timed out after 30000ms.
```

For the failover case I can force a refresh of the connection by performing a second AZ failover. In this case, the DNS record is ultimately updated to the original value, the service is started again, and the connections are properly terminated and new connections are being created. The connections that were in limbo now create exceptions

```
java.sql.SQLTransientConnectionException: HikariPool - 
    Connection is not available, request timed out after 690783ms.
```

which is obviously much longer than the configured value.

I'm lucky that I found this issue because this pointed me in the right direction (setting socket timeouts at the driver level) but this is still very surprising to me and I would still consider the current behavior a bug. I understand that this might not be fixed for 2.x series but it should be verified whether it's still a problem in 3.x.

**Comment by brettwooldridge** [comment]
See https://github.com/brettwooldridge/HikariCP/wiki/Rapid-Recovery

**Comment by musiKk** [comment]
I'm not sure I agree but maybe I don't understand it fully.

I get the point about connections borrowed by the application but in this case `HikariPool#getConnection` blocks for longer than specified in `connectionTimeout`. Even if all N connections of the pool are currently handed out to the application and blocked, the "N+1"th request should time out after `connectionTimeout` elapsed. It doesn't and the second log message in the comment above demonstrates that.

**Comment by brettwooldridge** [comment]
@musiKk A couple of things to note here.  This issue is reported against a version of the library that is 2 years old.  There was a bug fixed in v3.3.0 that according to the change log says:
> Revert change where Connection.isClosed() was called outside of setNetworkTimeout() block, opening vulnerability to unacknowledged TCP traffic.

It could very well be this issue.  It could just as easily be a driver bug where ``setNetworkTimeout()`` is not properly applied, such as during connection establishment.  Which ever the case, and in order to guard against both undiscovered bugs or future regressions, following the recommendations of the Rapid Recovery guide protects against all of them.

**Comment by musiKk** [comment]
Not arguing against what the recovery guide says. It provides very useful information! 👍

Regarding the fix in the newer version: Our application runs in an old stack. Updating the library to HikariCP 3 was unfortunately not possible. Good to know that this is potentially fixed now.

Thank you very much for being so responsive and providing your insights!
